guides/hack/15-asynchronous-operations/14-concurrent.md

```diff
@@ -34,3 +34,25 @@ concurrent {
   $result[] = await z_async();
 }
 ```
+
+## Exceptions
+
+If any statement in a `concurrent` block throws, evaluation of the other statements is undefined. They may have already run, they may never run, or they may still be executing. The concurrent block will throw an arbitrary single exception.
```

Comment:
This is only partly correct. When a concurrent block has completed, all the await operations in it will have completed in some way (either by throwing or returning a value). If any of them threw, then there are no guarantees about which (if any) returned values were assigned and which exception will be propagated if more than one of them threw. For example if you have:
```
$x = 0;
try {
  concurrent {
    $x = await async { return 1; };
    await async { throw new Exception('foo'); }
    await async { throw new Exception('bar'); }
  }
} catch (Exception $e) {
  var_dump($x, $e->getMessage());
}
```
Then it is explicitly undefined whether var_dump will see `$x === 0` or `$x === 1` and whether the message will be 'foo' or 'bar'.

Reply:
So if the runtime decides to order the awaitables in the order 3,2,1, all awaitables will execute (so all side effects of all the awaitables happen). Even though exception foo happened after exception bar, the runtime could decide to propagate foo instead. The assignment to x may or may not happen.

Am I on the right track here?
Going with the following example:
```HACK
$result = vec[];
concurrent {
  $result[] = await x_async();
  $result[] = await y_async();
}
```

If no exception is thrown, result will be [x, y], if x throws, result may be [] or [y], if y throws result maybe [] or [x].

If this is the case, the heading above could use a *if no exceptions occur*. If this is well defined, because the mutated the same variable, then it can stay as is.

Reply:
That's almost right. As it says above the code snippet in the doc we currently have, assignment of awaited results is guaranteed to happen in order, so we can't get a result that's [y] (because we would have had to check the result of x first and that would have propagated the exception from x), but $result may be any of [] (possible if either x or y or both throw) or [x] (possible if y throws), and will be [x, y] if no exception is thrown.

In any case, all the side-effects that all the Awaitables produce before they return or throw will happen, and they will all happen before any results are assigned by the concurrent block, but the order they happen relative to each other is undefined.

It is recommended that programmers not rely on the values of any variables assigned to in the concurrent block in the event that an exception is thrown out of it. The order of assignments is guaranteed, so it's possible to reason about them, but code will be more readable and maintainable if you don't.